kyo-prelude/native/src/main/scala/kyo/kernel/TracePool.scala

```diff
@@ -0,0 +1,49 @@
+package kyo.kernel
+
+import internal.*
+import java.util.Arrays
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicInteger
+import kyo.discard
+import scala.annotation.tailrec
+
+private[kernel] object TracePool:
+    inline def globalCapacity: Int = 10240
+    inline def localCapacity: Int  = 32
+
+    private val global = new ConcurrentLinkedQueue[Trace]()
+
+    abstract class Local:
+        final private val pool = new Array[Trace](localCapacity)
+        final private var size = 0
+
+        final def borrow(): Trace =
+            if size == 0 then
+                val trace = global.poll()
+                if trace != null then trace else Trace.init
```

Comment:
Also less efficient than the JVM version because it doesn't replenish from the global pool in batches.